server/server.go

```diff
@@ -6,53 +6,299 @@ package server
+	persistedIndex int64
 
-	queue *queue.Queue
+	polling       polling
+	notifications notifications
+
+	close *sync.Once
 
 	pb.UnimplementedProducerServer
 }
 
-// PublishEvents is the server implementation of the gRPC PublishEvents call
-func (serv shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
-	results := []*messages.EventResult{}
-	for _, evt := range req.Events {
-		serv.logger.Infof("Got event %s: %#v", evt.EventId, evt.Fields.Data)
-		err := serv.queue.Publish(evt)
-		if err != nil {
-			// If we couldn't accept any events, return the error directly. Otherwise,
-			// just return success on however many events we were able to handle.
-			if len(results) == 0 {
-				return nil, err
+type polling struct {
+	ctx     context.Context
+	stop    func()
+	stopped chan struct{}
+}
+
+// NewShipperServer creates a new server instance for handling gRPC endpoints.
+func NewShipperServer(q *queue.Queue, cfg ShipperServerConfig) (ShipperServer, error) {
+	if q == nil {
+		return nil, errors.New("queue cannot be nil")
+	}
+
+	err := cfg.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:   id.String(),
+		logger: logp.NewLogger("shipper-server"),
+		queue:  q,
+		cfg:    cfg,
+		polling: polling{
+			stopped: make(chan struct{}),
+		},
+		notifications: notifications{
+			subscribers: make(map[uuid.UUID]chan change),
+			mutex:       &sync.Mutex{},
+		},
+		close: &sync.Once{},
+	}
+
+	s.polling.ctx, s.polling.stop = context.WithCancel(context.Background())
+	s.startPolling()
+
+	return &s, nil
+}
+
+// GetAcceptedIndex atomically reads the accepted index
+func (serv *shipperServer) GetAcceptedIndex() int64 {
+	return atomic.LoadInt64(&serv.acceptedIndex)
+}
+
+// GetPersistedIndex atomically reads the persisted index
+func (serv *shipperServer) GetPersistedIndex() int64 {
+	return atomic.LoadInt64(&serv.persistedIndex)
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
+	resp := &messages.PublishReply{
+		Uuid: serv.uuid,
+	}
+
+	// the value in the request is optional
+	if req.Uuid != "" && req.Uuid != serv.uuid {
+		resp.AcceptedIndex = serv.GetAcceptedIndex()
+		resp.PersistedIndex = serv.GetPersistedIndex()
+		serv.logger.
+			With(
+				"expected", serv.uuid,
+				"actual", req.Uuid,
+			).
+			Debugf("shipper UUID does not match, all events rejected")
+
+		return resp, nil
+	}
+
+	for _, e := range req.Events {
+		err := serv.queue.Publish(e)
+		if err == nil {
+			resp.AcceptedIndex++
+			continue
+		}
+
+		log := serv.logger.
+			With(
+				"event_count", len(req.Events),
+				"accepted_count", resp.AcceptedCount,
+			)
+
+		if errors.Is(err, queue.ErrQueueIsFull) {
```

Comment:
I would if we should give some indication that the queue was full back to API clients in the response? 

Not as an error but just as a way for them to observe the state so they have a way to use a smarter retry strategy when this happens than just retrying immediately?

I think in the original shipper proposal we wanted to have the Publish call block until the events were accepted as a way of backpressuring the producer. Could we just wait on a channel from the queue until the condition clears or the request timeout expires? The next request isn't going to succeed until the queue actually has space to accept events.

@faec what do you think?

Reply:
> I would if we should give some indication that the queue was full back to API clients in the response?

Well, client would receive `AcceptedCount = 0` in this case. It should be enough.

> I think in the original shipper proposal we wanted to have the Publish call block until the events were accepted as a way of backpressuring the producer. Could we just wait on a channel from the queue until the condition clears or the request timeout expires? The next request isn't going to succeed until the queue actually has space to accept events.

I would implement something like this as a streaming endpoint and leave this as it is.

Reply:
I don't think we need to give extra feedback -- anytime less than our full event list was accepted, we know the queue is (instantaneously) full, and inputs should probably do the usual exponential backoff when that happens.

The scenarios I can think of where blocking would matter are:
- two inputs are competing for an almost-full queue and one has significantly smaller events than the other -- in that case we could get starvation of the larger events since there wouldn't be a blocking mechanism to keep its place in line. This is probably not a huge deal either way since it only arises when things are already very congested?
- An input trying to send many large events to a mostly-full queue -- in that case things work fine either way, but if the input is trying to send a large block of events it could end up sending the same large RPCs many times on an already-overloaded system. This is probably hot that bad if the input is a little more clever about exponential backoff -- say, apply backoff on the API calls and reduce batch size to whatever was accepted last time?

What if we always block on publish (up to timeouts, maybe) but only until at least one event is accepted? I.e. block enough to guarantee some progress but don't wait on more than that if the queue is full?

Reply:
The second case is mostly what I am worried about. Beyond existing beats, we won't have any control over what inputs do when they get the `accepted == 0` response. We can recommend exponential backoff but we can't force that to happen.

I would like to simplify the client behaviour has much as we can, where we don't choose each client to have to pick a backoff implementation and hope it is reasonable.

> What if we always block on publish (up to timeouts, maybe) but only until at least one event is accepted? I.e. block enough to guarantee some progress but don't wait on more than that if the queue is full?

I think this sounds reasonable as it makes sure a poorly implemented client doesn't do the equivalent of busy-waiting on the shipper queue.

Reply:
I still like the idea of having `PublishEvents` block if the queue is full, and unblocking as along as at least one event is accepted. 

This gives us a nice way to backpressure the inputs, which I prefer ove having multiple input processes pointlessly polling on the shipper in various backoff loops. It at least eliminates the latency of the backoff delay from publishing the next event when the queue next has space.

Reply:
@cmacknz it's not clear to me what you mean by "block if the queue is full", let's operate with more practical examples here:

## Example 1. The queue filled up during the request (logic already in place)

1. The client requests to publish 50 events
2. The queue can accept only 5 and we get "queue is full" error back from publishing the 6th
3. The server replies to the client with the current `AcceptedCount` and `AcceptedIndex` values partially accepting the event batch

## Example 2. The queue is full from the beginning (needs to be implemented)

1. The client requests to publish 50 events
2. When publishing the first event, the queue returns "queue is full" error and cannot accept even a single event
3. The server goes into an infinite loop (can be interrupted by the request context or a timeout) with `x` ms intervals trying to publish the first event to the queue – this blocks the client.
4. Once at least one event is accepted by the queue this switches to the behaviour in `Example 1`.

So, this is sort of an on-server backoff implementation. 

## My concerns

1. For `Example 2` it's the lack of observability. From the client's perspective, this will look like just a slow server. The client would not know about the reason why the request is so slow.
2. A lot of clients will compete for the queue to become empty and this solution does not really solve anything regarding this point. We could, of course, implement a very sophisticated ordering system that will do "first come – first served" for the queue publishing requests but it will complicate things and might make certain requests extremely slow.

Reply:
Thanks for the examples, I'm going to move this conversation into an issue so we can continue it outside the PR.